app/src/main/java/supermarket/model/Offer.kt

```diff
@@ -1,32 +1,78 @@
 package supermarket.model
 
+import java.lang.IllegalArgumentException
+import java.lang.IllegalStateException
+
 sealed class Offer {
     abstract fun getDiscount(
         shoppingCart: ShoppingCart,
         catalog: SupermarketCatalog
     ): Discount?
+
+    protected fun multiplierStringOrEmpty(numDiscounts: Int) =
+        if (numDiscounts > 1) {
+            " x${numDiscounts}"
+        } else {
+            ""
+        }
 }
 
-abstract class SingleProductOffer(val product: Product): Offer()
+class BundleOffer(
+    private val description: String,
+    private val products: List<Product>,
+    private val discountPrice: Double
+) : Offer()
+{
+    init {
+        if (products.isEmpty()) throw IllegalArgumentException("Must have some products in bundle")
+    }
 
-class PercentageOffer(product: Product, val percentOff: Double) : SingleProductOffer(product) {
+    override fun getDiscount(shoppingCart: ShoppingCart, catalog: SupermarketCatalog): Discount? {
+        val quantityOfBundleProducts = products.mapNotNull { shoppingCart.quantityOf(it) }
+        return if (quantityOfBundleProducts.size == products.size) {
```

Comment:
This works as long as every bundle only allows one of each product.  Imagine you had the "S'mores bundle" that provides a discount when you have 2 graham crackers, 1 marshmallow and 1 chocolate.  If you buy only one graham cracker though, this will apply the discount.

Reply:
Yep, this is a known limitation of this version. 😄  Support for Bundles with cardinality > 1 is TBD, and should be easy enough to add later with a `BundleSet` or some such class as an argument instead of a list of Products.